### ISSUE TITLE
chore: re-arrange sources to better reflect npm package and import paths

### ISSUE THREAD

**Issue by jeffbcross** [angular/angular#3713]
See [this comment](https://github.com/angular/angular/issues/3458#issuecomment-132391409) for details.

Essentially, the folder structure should be updated to look like this:

```
- modules/
--- angular2/
----- core/
----- forms/
----- http/...
----- router/...
----- web-worker/...
--- benchmarks/...
--- benchmarks_external/...
--- benchpress/...
--- examples/...
--- material2/...
--- rtts_assert/... (deprecated)
--- transforms/...
```

TODO:

 - [ ] Move `docs` outside of `angular2`.
 - [ ] Break `test_lib` and `mocks` into separate bundle-specific test folders, `http_test`, `core_test`, `router_test`.
 - [x] Add `Breaking Change` notes to commits
 - [x] Release ts2dart with updated paths https://github.com/angular/ts2dart/pull/266
 - [x] Remove superfluous top-level exports
 - [x] Add new public API exports to respective files/bundles.
 - [ ] Clean up the decorators spec to not leak between specs `angular2/test/core/util/decorators` spec "should not apply decorators from the prototype chain".
 - [x] Manually test npm version of benchpress after all changes (it depends on di)
 - [ ] Test angular.io to make sure correct paths are reflected
 - [ ] Maybe re-write all `angular2/angular2` imports to be more specific.
 - [x] Update `test` paths to reflect `src`
 - [x] Consolidate exports in src folders (to reduce redundancy between core and core_no_render exports).

**Comment by gkalpak** [comment]
It would be nice to be consistent with filename separators (e.g. `web-worker` vs `rtts_assert` (`-` vs `_`)).

**Comment by mhevery** [comment]
Rename `web-worker` to `web_worker`
I think we should also make `angular2/directives` and `angular2/pipes` because they could technically be outside of the core.

**Comment by jeffbcross** [comment]
K.

I made some progress on the transition on [this branch](https://github.com/jeffbcross/angular/tree/reorg) but have an issue with parse5Adapter not correctly setting DOM in time for application_spec. I'll have to revisit next week.

I also started on removing `src` and `test`, but will take care of that after moving everything around.

**Comment by jeffbcross** [comment]
I resolved blockers with test.unit.cjs and test.unit.js. Now on to fixing dart issues.

**Comment by jeffbcross** [comment]
FWIW, I proposed in-person to @IgorMinar, @tbosch and @rkirov that we not worry about flattening/merging `src` and `test` directories right now, as it would add significant complexity to our build process (moving folders for dart output to look like sources live in src/ and tests live in test/, as well as re-writing import paths). I think there was consensus of "yes, don't bother with this right now." I think we could make this change later without it being a breaking change, since import statements that include `src` should only be internal, not part of public API.

**Comment by jeffbcross** [comment]
Update: After updating ts2dart, there are still some issues I'm working through. This is the latest failing job: https://travis-ci.org/angular/angular/builds/77102478

Some problems: 
 * `@Inject` decorators being left in constructors where they shouldn't be. 
 * Some unused dart imports
 * http.d.ts `type` declarations being interpreted as `type-alias`, so [test](https://travis-ci.org/angular/angular/jobs/77102485) is faliling
 * Some decorators/compiler [unit tests](https://travis-ci.org/angular/angular/jobs/77102483) are failing because of mismatched number expectations.

**Comment by jeffbcross** [comment]
Some issues have been resolved.

 * `@Inject` was a matter of the decorator being imported instead of annotation.
* The http.d.ts problem was an existing problem that hasn't been tested. The problem has been temporarily fixed with an issue #3828 opened for long-term fix.

**Comment by jeffbcross** [comment]
I've merged the initial work of moving folders around, and am working on the rest of the TODOs in this issue.

**Comment by jeffbcross** [comment]
@mhevery in moving the default pipes outside of core, I assume I'll keep the `DEFAULT_PIPES_TOKEN` inside of core, bound to an empty list? The compiler relies on this binding so that it can flatten a View's pipes and the default pipes. So if a user wanted to use the default pipes, they would override the binding like so:

```typescript
bootstrap(MyComponent, [
  bind(DEFAULT_PIPES_TOKEN).toAlias(DEFAULT_PIPES)
]);
```

**Comment by mhevery** [comment]
@jeffbcross Pipes and Directives should be set up same way. Since directives are part of the core, I think pipes should be as well.

**Comment by jeffbcross** [comment]
Recapping VC with @mhevery:

 * Moving forms into core, and adding default bindings to root injector.
 * Keeping Pipes in core and keeping pipes' bindings in root injector.
 * Keeping directives in core bundle
 * angular2 bundle will include core, http, router concatenated, but will not include http and router bindings by default.

Bundles:
```
|- angular2.js
|--- core.js
|--- core_no_render.js
|--- http.js
|--- router.js
|--- web_worker.js
```

Plus testing bundles

**Comment by jeffbcross** [comment]
Closing this as the major work has been landed as of f14b212dc997a021d874ac0fc3e0e4d57a734fe4, and several follow-up issues have been filed:

 * Urgent: #4026 docs(typings): make typings match bundles 
 * Urgent: #4025 refactor: reorganize testing bundle and import
 * Urgent: #4023 refactor: include http and router in angular2.ts 
 * #4037 chore: move src/compiler/* to src/core/compiler/ 
 * #4036 discuss: remove angular2/bootstrap and angular2/render.
 * #4028 refactor: move angular2/lifecycle into angular2/core
 * #4027 refactor: deprecate angular2/angular2 import for dart
